### PR TITLE
Melt `buffer` down. 

### PR DESCRIPTION
**Pull request by andersio**
~~All hail the `ReplayState`. Just simplified how `replayLazily` initialises the replaying producer.~~
`replayLazily` now subsumes the implementation of `buffer`.

### CODE REVIEW

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -1695,26 +1682,27 @@ private final class DeallocationToken {
 	}
 }
 
-extension SignalProducer {
-	private static func bufferingProducer(upTo capacity: Int) -> (SignalProducer, Signal<Value, Error>.Observer) {
-		precondition(capacity >= 0, "Invalid capacity: \(capacity)")
+private struct ReplayState<Value, Error: ErrorProtocol> {
+	let producer: SignalProducer<Value, Error>
+	let observer: SignalProducer<Value, Error>.ProducedSignal.Observer
```
Comment (mdiep) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This seems a bit odd to me for a few reasons:

1. `ReplayState` is used with `Atomic` by `replayLazily` but itself uses an `Atomic<BufferState<Value, Error>>`.
2. `BufferState` is only used inside `ReplayState`.
3. `ReplayState` has a producer and observer, which means that it's more than just state information

How would you feel about attempting to unify this code a bit more?

Comment (andersio) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Sure. 👍

**Comment by andersio** [comment]
(Not ready yet)

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 2]
```diff
@@ -1745,75 +1730,42 @@ extension SignalProducer {
...
-			return
-		}
-
-		values.append(value)
-
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeSubrange(0..<overflow)
-		}
-	}
```
Comment (mdiep) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This seems like it's still useful to have as a `mutating func`. Keeping the state manipulation here makes it a little easier to follow the higher-level threading logic in the producer.

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 3]
```diff
@@ -1640,81 +1638,68 @@ extension SignalProducerProtocol {
...
-
-		let lock = Lock()
-		lock.name = "org.reactivecocoa.ReactiveCocoa.SignalProducer.replayLazily"
-
 		// This will go "out of scope" when the returned `SignalProducer` goes
 		// out of scope. This lets us know when we're supposed to dispose the
 		// underlying producer. This is necessary because `struct`s don't have
 		// `deinit`.
-		let token = DeallocationToken()
+		let lifetime = DeallocationToken()
```
Comment (mdiep) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This should be able to use a `Lifetime` now?

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 4]
```diff
@@ -1674,172 +1672,175 @@ extension SignalProducerProtocol {
...
+			// attachments.
+			terminationEvent = event
+			observers = nil
 		}
 	}
+
+	/// Broadcast the event to all observers.
+	///
+	/// - parameter:
+	///   - event: The event to be broadcasted.
```
Comment (andersio) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Oops. Spelling mistake. Will fix ASAP.

**Comment by andersio** [comment]
An experiment of encapsulating all state manipulations.

File: ReactiveCocoa/Swift/Observer.swift [hunk 5]
```diff
@@ -26,7 +26,7 @@ public protocol ObserverProtocol {
 
 /// An Observer is a simple wrapper around a function which can receive Events
 /// (typically from a Signal).
-public struct Observer<Value, Error: ErrorProtocol> {
+public class Observer<Value, Error: ErrorProtocol> {
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Observer.swift]: +1 I think this is more correct.
It should be `final` though!

File: ReactiveCocoa/Swift/Observer.swift [hunk 6]
```diff
@@ -70,6 +70,16 @@ public struct Observer<Value, Error: ErrorProtocol> {
 	}
 }
 
+extension Observer: Hashable {
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Observer.swift]: Hmm I don't think it makes sense to give identity to `Observer` though :/

File: ReactiveCocoa/Swift/Observer.swift (hunk 5 above)
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Observer.swift]: Bump

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 7]
```diff
@@ -1674,172 +1672,187 @@ extension SignalProducerProtocol {
...
-				if let value = event.value {
-					state.add(value, upTo: capacity)
-				} else {
-					// Disconnect all observers and prevent future
-					// attachments.
-					state.terminationEvent = event
-					state.observers = nil
+					// Terminate the replay loop.
+					return
+				} catch ReplayError<Value>.pending(let values) {
```
Comment (andersio) [inline ReactiveCocoa/Swift/SignalProducer.swift]: The compiler somehow does not permit `catch let ReplayError<Value>.pending(values)`.

**Comment by andersio** [comment]
Please note that the semantic has slightly changed: the underlying producer is captured by the bootstrap disposable, so it would be released after it is being started. In the original implementation, it was captured by the replayed producer.

**Comment by mdiep** [comment]
I merged the `RAC5` branch, so this will need to be reopened against `master`. Sorry!

**Pull request closed without merging**